[DOC] mm2 topic offset synchronization (#4473)
strimzi/strimzi-kafka-operator: documentation/modules/configuring/con-config-mirrormaker2-replication.adoc

MirrorMaker 2.0 is based on the Kafka Connect framework, _connectors_ managing the transfer of data between clusters.
A MirrorMaker 2.0 `MirrorSourceConnector` replicates topics from a source cluster to a target cluster.

By default, a check for new topics in the source cluster is made every 10 minutes.
You can change the frequency by adding `refresh.topics.interval.seconds` to the source connector configuration of the `KafkaMirrorMaker2` resource.
However, increasing the frequency of the operation might affect overall performance.

> 

Reviewer: I have limited knowledge of MM2, but the new paragraph seems to interrupt the flow of the original introduction (which is great). It talks about the `KafkaMirrorMaker2` resource very early on.

Perhaps it could be positioned after line 25, or even in the "Cluster configuration" section.

Author: Fair point. I've moved to after the image

MirrorMaker 2.0 uses its `MirrorCheckpointConnector` to emit _checkpoints_ for offset tracking.

== Synchronizing consumer group offsets

> The `__consumer_offsets` topic stores information on committed offsets, according to consumer group.

Reviewer 2: ```suggestion
The `__consumer_offsets` topic stores information on committed offsets, for each consumer group.
```

> Offset synchronization periodically transfers the consumer offsets for the consumer groups of a source cluster into the consumer offsets topic of a target cluster.

Reviewer: "Offset synchronization" is a MM2 concept, correct?

Author: Yes

Offset synchronization is particularly useful in an _active/passive_ configuration.

> Should the active cluster go down, consumer applications can switch to the passive (standby) cluster and pick up from the last offset position.

Reviewer: ```suggestion
If the active cluster goes down, consumer applications can switch to the passive (standby) cluster and pick up from the last offset position.
```

> Should the active cluster go down, consumer applications can switch to the passive (standby) cluster and pick up from the last offset position.

Reviewer 2: ```suggestion
Should the active cluster go down, consumer applications can switch to the passive (standby) cluster and pick up from the last transferred offset position.
```

To use topic offset synchronization:

* Enable the synchronization by adding `sync.group.offsets.enabled` to the checkpoint connector configuration of the `KafkaMirrorMaker2` resource,
and setting the property to `true`. Synchronization is disabled by default.
* Add the `IdentityReplicationPolicy` to the source connector configuration so that topics in the target cluster retain their original names.

> For topic offset synchronization to work, consumer groups in the source cluster cannot be active in the target cluster.

Reviewer: ```suggestion
For topic offset synchronization to work, consumer groups in the source cluster cannot also be active in the target cluster.
```

> For topic offset synchronization to work, consumer groups in the source cluster cannot be active in the target cluster.

Reviewer 2: I think we can phrase this clearer, since AFAIU the problem is just about colliding group ids, rather than use the of consumer groups _per se_.

```suggestion
For topic offset synchronization to work, consumer groups in the target cluster cannot use the same ids as groups in  the source cluster.
```

> Consequently, if you are using an _active/active_ configuration that does not support or require offset synchronization, do not enable topic offset synchronization.

Reviewer 3: I'm not sure I understand this sentence. In general, you can use offset synchronization also in active-active clusters. And since you anyway cannot easily use a single consumer group over multiple clusters, this might not be a big limitation I think. So maybe we can remove this sentence?

Author: Removed

If enabled, the synchronization of offsets from the source cluster is made periodically.
You can change the frequency by adding `sync.group.offsets.interval.seconds` and `emit.checkpoints.interval.seconds` to the checkpoint connector configuration.
The properties specify the frequency in seconds that the consumer group offsets are synchronized, and the frequency of checkpoints emitted for offset tracking.

> The default for both properties is 60 seconds.

Reviewer 3: Should this mention also the `refresh.groups.interval.seconds` interval? (default 10 minutes)

Reviewer 2: We should also be explicit that because this is time based it means that consumers which switch to the passive on failover will process some (hopefully small) number of duplicates.

Author: Updated to:

> If enabled, the synchronization of offsets from the source cluster is made periodically.
> You can change the frequency by adding `sync.group.offsets.interval.seconds` and `emit.checkpoints.interval.seconds` to the checkpoint connector configuration.
> The properties specify the frequency in seconds that the consumer group offsets are synchronized, and the frequency of checkpoints emitted for offset tracking.
> The default for both properties is 60 seconds.
> You might also want to change the frequency of checks for new consumer groups using the `refresh.groups.interval.seconds` property, which is performed every 10 minutes by default.
> 
> Because the synchronization is time-based, any switchover by consumers to a passive cluster will likely result in some duplication of messages.

Also added `refresh.groups.interval.seconds` to example config

If the active cluster goes down, consumer applications can switch to the passive (standby) cluster and pick up from the last transferred offset position.

To use topic offset synchronization:

* Enable the synchronization by adding `sync.group.offsets.enabled` to the checkpoint connector configuration of the `KafkaMirrorMaker2` resource,
and setting the property to `true`. Synchronization is disabled by default.
* Add the `IdentityReplicationPolicy` to the source and checkpoint connector configuration so that topics in the target cluster retain their original names.

For topic offset synchronization to work, consumer groups in the target cluster cannot use the same ids as groups in the source cluster.

If enabled, the synchronization of offsets from the source cluster is made periodically.
You can change the frequency by adding `sync.group.offsets.interval.seconds` and `emit.checkpoints.interval.seconds` to the checkpoint connector configuration.
The properties specify the frequency in seconds that the consumer group offsets are synchronized, and the frequency of checkpoints emitted for offset tracking.
The default for both properties is 60 seconds.

> You might also want to change the frequency of checks for new consumer groups using the `refresh.groups.interval.seconds` property, which is performed every 10 minutes by default.

Reviewer 3: TBH, I'm not sure we should suggest the change. We do not really know what impact these have on performance etc. So I think saying _You can also change ..._ would be better.

Author: Changed to: _You can also change the frequency of checks for new consumer groups using the `refresh.groups.interval.seconds` property, which is performed every 10 minutes by default._
